### ISSUE TITLE
$scope.form.$valid is returning false even it is true.

### ISSUE THREAD

**Issue by esh-cherukuri** [ghiscoding/angular-validation#79]
Hi,
Some times $scope.myform.$valid is returning false even it is valid. i don't understand why it is happening please help me out.

Thanks,
Rajesh.

**Comment by ghiscoding** [comment]
Unless you provide a Plunker that I can troubleshoot with, I can't really help you.

**Comment by esh-cherukuri** [comment]
Hi,
I had created a plunker. you can check it here.http://plnkr.co/edit/y6HqYhxoqWSlj8aWqEz7?p=preview
try entering single character i will alert you the onchange  $valid value. Some times it is displaying false and some times it is displaying true.

**Comment by ghiscoding** [comment]
I see the effect as you said, but it seems to happen only within the Controller. It's probably a timing issue, there's a few timer and also the $digest that could affect, so I'll see what I can do.

On the other side or testing, disabling the button with `form1.$invalid` inside the html code is always working correctly.
`<button type="submit" name="btn_ngDisabled" class="btn btn-primary" ng-disabled="form1.$invalid" >Save</button>`

I'm also curious, why are you looking at `form1.$valid` inside the Controller? I usually have a call to `checkFormValidity()` with the `validationService` to do just that. But actually, I'll have to look at that function too since it seems to have the same issue there too.

**Comment by esh-cherukuri** [comment]
Even it is not working properly with checkFormValidity(). When there is no data it is returning true, when i enter 1 character it is returning false and then true. Please check the updated plunker.
http://plnkr.co/edit/y6HqYhxoqWSlj8aWqEz7?p=preview

**Comment by ghiscoding** [comment]
I know since I've already tried it. I think the problem is around the fact that you use `ngChange`, there seem to be a timing issue, possibly with the `$digest`. What is the particular reason that you use `ngChange`?

**Comment by esh-cherukuri** [comment]
When there is a change in input field i need to enable save and cancel buttons (by default they are disabled). if user clicks on save or cancel button i need to call corresponding function and again i need to disable the buttons until there is a change in input fields.

**Comment by ghiscoding** [comment]
Ok so I found what the problem is. It might not be clearly visible to you but the conflict is simply by the fact that I use a timer of 1sec (by default) to even start doing validation, but in your controller you use `ngChange` right away without waiting for my validation to even start. 

If you add a timer, which is over my 1sec timer, into your `onChange()`, function it does work (even if that doesn't look like nice code)
```javascript
$scope.onChange = function(){
   $timeout(function(){
     $scope.result = $scope.form1.$valid;
     $scope.checkFormValidity = new validationService().checkFormValidity($scope.form1);  
   },1100);
```

On the other hand if you try removing the timer by setting the `debounce` to 0, you still need to use a `$timeout` simply because using `form1.$valid` inside the Controller is just too fast and too early in the `$digest` to get correct result. For example, this would work:
```javascript
 var vs = new validationService({ debounce: 0 });
 
 $scope.onChange = function(){
   $timeout(function(){
     $scope.result = $scope.form1.$valid;
     $scope.checkFormValidity = new validationService().checkFormValidity($scope.form1);  
   },5);
```
but if you try making it under 5ms, you'll have timing issue again. 

So let's resume all that, the way you want to check the `form1.$valid` in the Controller is simply too fast and too early in the `$digest` to work correctly with my Directive. You can counter that problem with a `$timeout` even if ugly, it does work. I'm not sure how else to fix your current implementation. Hope that helps... if you have other ideas then say it.

**Comment by ghiscoding** [comment]
Actually this makes me think that there might be another option but that would require code change on my side. Inside my Directive, I use a function named `attemptToValidate()` which return a promise. I use this for other purposes, but if you want, I could possibly create an attribute for a callback when the promise is finished. 

Please note that this is just an idea, it might not work. But let's take an example of this possibly.
```html
<input name="firstname" ng-model="firstName" validation="required:alt=First name is required"
validation-callback="onChange">
```
So basically you would use a new attribute of `validation-callback` instead of `ng-change`, that could possibly work since I already deal internally with a promise.

**Comment by esh-cherukuri** [comment]
This would be greatful with out using above ugly code. can you please test it by making change on your side by adding new attribute. if that works it will be very helpful to me.

**Comment by ghiscoding** [comment]
The concept works, I just tried it now. I'll have to look and see if I can add it to the Service as well. I'll have it done by the end of the week or maximum next week.

**Comment by esh-cherukuri** [comment]
ok Thank you.

**Comment by ghiscoding** [comment]
Added the `validation-callback` like I told you before. So you could use it this way
```html
<input name="firstname" ng-model="firstName" validation="required:alt=First name is required"
validation-callback="onChange()">
```

Pushed in latest version `1.4.14`, please see the [Wiki - Validation Callback](https://github.com/ghiscoding/angular-validation/wiki/Validation-Callback)

You could also look into the folder `more-examples/validationCallback` for a full demo. Enjoy it...

**Comment by ghiscoding** [comment]
If you like the project, you could also click on **Star** on top, that's all I ask... enjoy :)

**Comment by esh-cherukuri** [comment]
Thank you so much for your help... Its done i can give to multiple stars...:)
